File: util/android/src/main/java/com/dmi/util/db/DatabaseUpgrades.java
package com.dmi.util.db;

import android.annotation.SuppressLint;
import android.content.Context;
import android.database.SQLException;
import android.database.sqlite.SQLiteDatabase;

import com.google.common.base.Charsets;
import com.google.common.io.CharStreams;

import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.List;

import static java.lang.String.format;
import static java.util.Arrays.sort;

public abstract class DatabaseUpgrades {
    public static void upgradeDatabase(Context context, SQLiteDatabase database, String scriptsAssetsPath) throws DowngradeException {
        database.beginTransaction();
        try {
            NewScripts newScripts = getNewScripts(context, database.getVersion(), scriptsAssetsPath);
            if (newScripts.version < database.getVersion()) {
                throw new DowngradeException(
                        format("Database cannot be downgraded. Database version: %s, scripts version: %s",
                               database.getVersion(),
                               newScripts.version)
                );
            }
            applyScripts(context, database, newScripts.paths);
            database.setVersion(newScripts.version);
            database.setTransactionSuccessful();
        } catch (IOException | SQLException e) {
            throw new RuntimeException("Upgrade database error", e);
        } finally {
            database.endTransaction();
        }
    }

    private static NewScripts getNewScripts(Context context, int currentVersion, String scriptsAssetsPath) throws IOException {
        NewScripts newScripts = new NewScripts();
        String[] scripts = context.getAssets().list(scriptsAssetsPath);
        sort(scripts);
        for (String script : scripts) {
            if (script.endsWith(".sql")) {
                String scriptName = script.substring(0, script.indexOf('.'));
                int scriptVersion = Integer.parseInt(scriptName);
                if (scriptVersion > currentVersion) {
                    newScripts.paths.add(scriptsAssetsPath + '/' + script);
                }
                newScripts.version = scriptVersion;
            }
        }
        return newScripts;
    }

    private static void applyScripts(Context context, SQLiteDatabase database, List<String> scriptPaths) throws IOException {
        for (String scriptPath : scriptPaths) {
            for (String sqlCommand : readSQLCommands(context, scriptPath)) {
                database.execSQL(sqlCommand);
            }
        }
    }

    @SuppressLint("NewApi")
    private static List<String> readSQLCommands(Context context, String scriptPath) throws IOException {
        try (InputStream is = context.getAssets().open(scriptPath)) {
            return CharStreams.readLines(new InputStreamReader(is, Charsets.UTF_8));
        }
    }

    private static class NewScripts {
        public final List<String> paths = new ArrayList<>();
        public int version;
    }

    public static class DowngradeException extends Exception {
        public DowngradeException(String detailMessage) {
            super(detailMessage);
        }
    }
}

File: util/util-android/src/main/kotlin/com/dmi/util/db/DatabaseUpgrades.kt
package com.dmi.util.db

import android.content.Context
import android.database.SQLException
import android.database.sqlite.SQLiteDatabase
import com.google.common.base.Charsets
import com.google.common.io.CharStreams.readLines
import java.io.IOException
import java.io.InputStreamReader
import java.lang.Integer.parseInt
import java.util.*
import java.util.Arrays.sort

fun upgradeDatabase(context: Context, db: SQLiteDatabase, scriptsAssetsPath: String) {
    db.beginTransaction()
    try {
        val newScripts = findNewScripts(context, db.version, scriptsAssetsPath)
        if (newScripts.version < db.version) {
            throw DatabaseDowngradeException(
                    "Database cannot be downgraded. Database version: ${db.version}, scripts version: ${newScripts.version}"
            )
        }
        applyScripts(context, db, newScripts.paths)
        db.version = newScripts.version
        db.setTransactionSuccessful()
    } catch (e: IOException) {
        throw RuntimeException("Upgrade database error", e)
    } catch (e: SQLException) {
        throw RuntimeException("Upgrade database error", e)
    } finally {
        db.endTransaction()
    }
}

private fun findNewScripts(context: Context, currentVersion: Int, scriptsAssetsPath: String): NewScripts {
    val newScripts = NewScripts()
    val scripts = context.assets.list(scriptsAssetsPath)
    sort(scripts)
    for (script in scripts) {
        if (script.endsWith(".sql")) {
            val name = script.substring(0, script.indexOf('.'))
            val version = parseInt(name)
            if (version > currentVersion)
                newScripts.paths.add("$scriptsAssetsPath/$script")
            newScripts.version = version
        }
    }
    return newScripts
}

private fun applyScripts(context: Context, database: SQLiteDatabase, scriptPaths: List<String>) {
    for (scriptPath in scriptPaths) {
        for (sqlCommand in readSQLCommands(context, scriptPath)) {
            database.execSQL(sqlCommand)
        }
    }
}

private fun readSQLCommands(context: Context, scriptPath: String): List<String> {
    context.assets.open(scriptPath).use { stream ->
        return readLines(InputStreamReader(stream, Charsets.UTF_8))
    }
}

private class NewScripts {
    val paths: MutableList<String> = ArrayList()
    var version: Int = 0
}

class DatabaseDowngradeException(detailMessage: String) : Exception(detailMessage)